core/src/multi_proof.rs

```diff
@@ -0,0 +1,548 @@
+// and the siblings themselves.
+#[derive(Debug, Clone)]
+struct SubPathProof {
+    terminal: PathProofTerminal,
+    depth: u8,
+    inner_siblings: Vec<Node>,
+}
+
+/// A proof of multiple path through the trie.
+#[derive(Debug, Clone)]
+pub struct MultiProof {
+    // All subpaths related to a single terminal node
+    sub_paths: Vec<SubPathProof>,
+    // Vector containing the minimum number of nodes required
+    // to reconstruct all other nodes later.
+    external_siblings: Vec<Node>,
+}
+
+impl MultiProof {
+    /// Construct a MultiProof from a vector of *ordered* PathProof
+    pub fn from_path_proofs(path_proofs: Vec<PathProof>) -> Self {
+        // A multi-proof can be viewed by associating each terminal node
+        // with its first n uniquely related siblings from its path proof,
+        // followed by all necessary siblings that are not derivable from
+        // the previously mentioned siblings.
+        let mut sub_paths: Vec<SubPathProof> = vec![];
+        let mut external_siblings: Vec<Node> = vec![];
+
+        // The goal is to traverse the entire tree
+        // formed by all path proofs and only collect the siblings
+        // that cannot be reconstructed in the future
+        //
+        // The traversal does not occur on the siblings themselves,
+        // but on the ordered set of key_paths within PathProofs.
+        //
+        // For example, take two key_paths starting from the first bit.
+        // If two keys share the same bit at index i, they will have the same sibling at index i.
+        // If the bit at index i differs, their paths diverge, and no sibling is needed at that index
+        // because it can be reconstructed from siblings collected later in the other key.
+        // When a differing bit is found, each key_path needs separate scanning.
+        //
+        // When there is a single key_path, all siblings are necessary.
+        //
+        // When there are more than two key_paths, the same logic is applied,
+        // but if two key_paths have different bits at an index, the key_paths are split
+        // based on the value at that index.
+        //
+        // If all key_paths share a bit at an index, that sibling is required;
+        // if at least two key_paths differ, no sibling is needed, but the key_paths must be divided
+        // based on having bit 0 or 1 at that index.
+        //
+        // Iterate this algorithm on the bisection to determine the minimum necessary siblings.
+        //
+        // `external_siblings` whill follow this structure for each bicection:
+        // |commons siblings| ext siblings in the right bisection | ext siblings in the left bisection |
+
+        // it represent a slice of the vector of key_paths with the notion of
+        // path_index, at which bit in the key_path that slice is pointing at
+        struct TraverseInfo {
+            path_index: usize,
+            path_slice_lower: usize,
+            path_slice_upper: usize,
```

Comment:
it seems like this is an inclusive bound?

Reply:
Yes, a comment would definitely help. I will add it